File: Chapter 1 Arrays and String/URLify.py
class Solution:

    def __init__(self):
        pass

    def URLify(self, s,trueSize): #s is a list of characters, size is the true size
        p2 = trueSize-1 #the pointer to the new string
        p1 = trueSize-1 #the pointer to the origional string

        while(s[p1] == ' '):
            p1 -= 1

        while(p1 >= 0): #keep going through each of the original characters til we reach the end
            if(s[p1] != ' '): 
                s[p2] = s[p1]
                p2 -= 1
            else: #if we see a space replace it with %20
                s[p2] = '0'
                s[p2-1] = '2'
                s[p2-2] = '%'
                p2 -= 3
            p1 -= 1

File: Chapter 1 Arrays and String/URLify.java
public class URLify{
    public static void main(String[] args){
        Solution soln = new Solution();
        char[][] tests = {{'M','r',' ','J','o','h','n',' ','S','m','i','t','h',' ',' ',' ',' '}};
        for(char[] test : tests){
            soln.Url(test,13);
            String s = new String(test);
            System.out.println(s);
        }
    }
}

class Solution{
    public void Url(char[] str,int trueSize){
        int trueIndex = trueSize-1;
        int index = str.length-1;
        while(trueIndex >= 0){
            if(str[trueIndex] != ' '){
                str[index] = str[trueIndex];
                index -= 1;
            }
            else{
                str[index] = '0';
                str[index-1] = '2';
                str[index-2] = '%';
                index -= 3;
            }
            trueIndex -= 1;
        }
    }
}